### DISCUSSION TITLE
`defineRouteMeta({})` compiler macro

### DISCUSSION THREAD

**Discussion by danielroe** [nuxt/nuxt#18012] [Announcements]
# Background

There are a number of use cases where we need to get information from _pages_.

In some cases we care about the route, not the page component specifically. For example, a child component might decide we need to render with the 'admin' layout - this is a global value. In other cases, we care specifically about the value from the page itself (e.g. a parent component might not have a key, but a child component might have a key). In these cases, we want to tie the value specifically to the component data itself.

Here are some sample use cases.

## Layout information

> **Route-level**

We need to be able to define the layout at the route level. It therefore makes sense to define this on route meta rather than directly on the component.

See https://github.com/nuxt/framework/issues/2074 and https://github.com/nuxt/framework/issues/782

## Transition and keepAlive props

> **Route-level**

As above, this should be defined at route level. If we set transition props on a child route, we want this to be accesible in `NuxtPage` so it can be set correctly on the global `<transition>` element.

Note that this data isn't serialisable. For example, we can pass functions as transition options.

## Middleware

> **Route-level**

This should also be defined at route level and needs to be accessible outside of `<RouterView>` so we can handle it within route guards.

See https://github.com/nuxt/framework/discussions/2082

## Key

> **Page-level**

This should be set specifically on the component that needs to be keyed, rather than on route meta. We might have a shared layout or parent route that can persist between routes, but the child needs to be keyed specifically. We don't need to know the route key unless actually showing the route, which means it's safe to store within the component options.

Note that this data isn't serialisable. For example, we can pass a function that takes a route object.

## User-defined metadata

> **Route-level**

Users may wish to define metadata like page title, description, etc. These should be global rather than specific to the component, and make sense to store in the route meta object as a whole. Users may wish to know all this metadata even when not displaying the component.

# Implementation

## Usage within script setup

We can use compiler macros to define this data within `<script setup>` and make it available within `router.mjs` (for route-level data) and within the component options (for page-level data).

Would allow usage like:

```ts
// this will be extracted out to the route metadata within router.mjs
defineRouteMeta({
  middleware: []
})

// this will be hoisted to the options of the component
definePageKey(route => route.path)
definePageKey('fixed')
```

## Usage with Options API

```js
export default {
  middleware: [],
  key: route => route.path,
  // or key: 'fixed'
}
```

We could alternatively support using the same compiler macro as above in the Options API.

**Comment by psycongaroo** [comment]
There'll be another folder in root where we can add middleware/guards?

**Comment by pi0** [comment]
I love them! This meta can be probably also used for page/route-level caching specification (#560) also replacing [router-extras-module](https://github.com/nuxt-community/router-extras-module) features.

We can also think about alternative sources to override and extend this meta including `nuxt.config` and `_meta.ts` files within `pages/` directory to apply a meta or guard to a group of routes.

Regarding naming, since we are using term of `pages`, maybe it makes more sense to be `definePageMeta`?

Also regarding key, `definePageKey`, it seems an odd exception from other usecases. Maybe can be removed from this RFC in favor or a generic micro like `defineComponentOption`?

**Comment by richardeschloss** [comment]
I like the ideas presented here, can you please also consider supporting the use case of `layout` accepting an object. This way, modules that want to provide their own custom layouts can easily do so without having to copy the layouts over to the layouts directory.

Something like:
```js 
import AwesomeLayout from '@nuxt/ui'
definePageMeta({
  layouts: AwesomeLayout // this could be a JSON object { render () { ... } }
})
```
Or better yet, in the same way components get auto-registered, can layouts also be auto-registered / auto-prefixed? So that if layouts are also in some other directory, they can be made available?

Currently, if I want to do something like this, I have to do (say, from a plugin):
```js
import AwesomeLayout from '@nuxt/ui'
import layouts from '#build/layouts'
layouts.awesomeLayout = AwesomeLayout // works (even though it's not documented, I've verified it works)
```
